src/module/actor/character/index.ts

```diff
@@ -1242,190 +1200,104 @@ class CharacterPF2e extends CreaturePF2e {
+            levels: classFeatSlots?.ancestry ?? [],
+        });
+        this.feats.createCategory({
+            id: "class",
+            label: "PF2E.FeatClassHeader",
+            featFilter: "classes-" + this.class?.slug,
+            supported: ["class"],
+            levels: classFeatSlots?.class ?? [],
+        });
+
+        const evenLevels = new Array(this.level)
+            .fill(0)
+            .map((_, idx) => idx)
+            .filter((idx) => idx && idx % 2 === 0);
 
-            mergeObject(this.featGroups, {
-                ancestry: { feats: mapFeatLevels(classItem.data.ancestryFeatLevels?.value, "ancestry") },
-                class: { feats: mapFeatLevels(classItem.data.classFeatLevels?.value, "class") },
-                skill: { feats: mapFeatLevels(classItem.data.skillFeatLevels?.value, "skill") },
-                general: { feats: mapFeatLevels(classItem.data.generalFeatLevels?.value, "general") },
+        // Add dual class if active
+        if (game.settings.get("pf2e", "dualClassVariant")) {
+            this.feats.createCategory({
+                id: "dualclass",
+                label: "PF2E.FeatDualClassHeader",
+                supported: ["class"],
+                levels: [1, ...evenLevels],
             });
         }
 
-        if (game.settings.get("pf2e", "ancestryParagonVariant")) {
-            this.featGroups.ancestry?.feats.unshift({
-                id: "ancestry-bonus",
-                level: 1,
-                grants: [],
+        // Add free archetype (if active)
+        if (game.settings.get("pf2e", "freeArchetypeVariant")) {
+            this.feats.createCategory({
+                id: "archetype",
+                label: "PF2E.FeatArchetypeHeader",
+                supported: ["class"],
+                levels: evenLevels,
             });
-            for (let level = 3; level <= this.level; level += 4) {
-                const index = (level + 1) / 2;
-                this.featGroups.ancestry?.feats.splice(index, 0, { id: `ancestry-${level}`, level, grants: [] });
-            }
         }
 
+        this.feats.createCategory({
+            id: "skill",
+            label: "PF2E.FeatSkillHeader",
+            supported: ["skill"],
+            levels: classFeatSlots?.skill ?? [],
+        });
+        this.feats.createCategory({
+            id: "general",
+            label: "PF2E.FeatGeneralHeader",
+            supported: ["general", "skill"],
+            levels: classFeatSlots?.general ?? [],
+        });
+
+        // Add campaign feats if enabled
+        if (game.settings.get("pf2e", "campaignFeats")) {
+            this.feats.createCategory({ id: "campaign", label: "PF2E.FeatCampaignHeader" });
+        }
+
+        // Add background skill feat slot
         const background = this.background;
         if (background && Object.keys(background.data.data.items).length > 0) {
-            this.featGroups.skill?.feats.unshift({
+            this.feats.get("skill")?.feats.unshift({
```

Comment:
Can the `get` method signature get overloaded to make ones like these guaranteed to return an array?

Reply:
We could, but if we do it might be wise to move the creation of the base categories to the constructor to guarantee this is the case. That would completely get rid of CharacterPF2e#prepareFeats() but that might be fine and probably a better result anyways.

Reply:
Creating the core categories in the constructor would definitely be a necessary step.

Reply:
Should be good to go, prepareFeats() is still there of course cause there's other things that do still need doing there right now.